src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1906,204 @@ MutableBigInteger sqrt() {
+     * <a href="https://inria.hal.science/inria-00072854/en/">  here</a> and
+     * <a href="https://www.researchgate.net/publication/220532560_A_proof_of_GMP_square_root">  here</a>
+     */
+    private MutableBigInteger[] sqrtRemZimmermann(int len, boolean needRemainder) {
+        if (len == 2) { // Base case
+            long x = ((value[offset] & LONG_MASK) << 32) | (value[offset + 1] & LONG_MASK);
+            long s = (long) Math.sqrt(x > 0 ? x : x + 0x1p64);
+            if (s > LONG_MASK || Long.compareUnsigned(x, s * s) < 0)
+                s--;
+
+            // Allocate sufficient space to hold the normalized final square root
+            MutableBigInteger sqrt = new MutableBigInteger(new int[(intLen + ((-intLen) & 3) + 1) >> 1]);
+
+            // Place the partial square root
+            sqrt.intLen = 1;
+            sqrt.value[0] = (int) s;
+
+            // The first invocation is never a base case, so the remainder is needed
+            return new MutableBigInteger[] { sqrt, new MutableBigInteger(x - s * s) };
+        }
+
+        // Recursive step (len >= 3)
+
+        // Normalize
+        /* For speed, the normalization is performed only "logically",
+         * which means that the contents of the input are not changed,
+         * but only its logical length, which is stored in the parameter len.
+         * The length is normalized to a mutiple of 4 because, in this way,
+         * the input can be always split in blocks of words without twiddling with bits.
+         */
+        final int limit = len;
+        final int excessInts = (-len) & 3;
+        len += excessInts; // len must be a multiple of 4
+
+        MutableBigInteger[] sr = sqrtRemZimmermann(len >> 1, true); // Recursive invocation
+
+        final int blockLen = len >> 2, blockBitLen = blockLen << 5;
+        MutableBigInteger dividend = sr[1];
+        dividend.leftShift(blockBitLen);
+        dividend.add(getBlockZimmermann(1, len, limit, blockLen));
+        MutableBigInteger twiceSqrt = new MutableBigInteger(sr[0]);
+        twiceSqrt.leftShift(1);
+        MutableBigInteger q = new MutableBigInteger();
+        MutableBigInteger u = dividend.divide(twiceSqrt, q);
+
+        MutableBigInteger sqrt = sr[0];
+        sqrt.leftShift(blockBitLen);
+        sqrt.add(q);
+
+        MutableBigInteger chunk = u;
+        chunk.leftShift(blockBitLen);
+        chunk.add(getBlockZimmermann(0, len, limit, blockLen));
+        BigInteger qBig = q.toBigInteger(); // Cast to BigInteger to use fast multiplication
+        MutableBigInteger qSqr = new MutableBigInteger(qBig.multiply(qBig).mag);
+
+        MutableBigInteger rem;
+        if (needRemainder) {
+            rem = chunk;
+            if (rem.subtract(qSqr) == -1) {
+                twiceSqrt = new MutableBigInteger(sqrt);
+                twiceSqrt.leftShift(1);
+
+                rem.add(ONE);
+                rem.subtract(twiceSqrt);
```

Comment:
Shouldn't these be `rem + twiceSqrt - 1`?

Reply:
No, because `rem` is unsigned, so to perform the correct computation you must first add the absolute value of the addends with the same sign, and the subtract those with opposite sign.

Reply:
Ah right, this is the semantics of `MBI.subtract`.